rust/vapid/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+    // problems.
+    let r_len = signature[r_off] as usize;
+    // calculate the offsets for the byte array data we want.
+    let s_off: usize = r_off + r_len + 2;
+    let s_len = signature[s_off] as usize;
+    let mut r_val = &signature[(r_off + 1)..(r_off + 1 + r_len)];
+    let mut s_val = &signature[(s_off + 1)..(s_off + 1 + s_len)];
+    // Strip the leading 0 if it's present.
+    if r_len == 33 && r_val[0] == 0 {
+        r_val = &r_val[1..];
+    }
+    if s_len == 33 && s_val[0] == 0 {
+        s_val = &s_val[1..];
+    }
+    // we now have the r and s byte arrays. Build the raw RS we need for the signature
+    // println!("r_val: ({}){:?}\ns_val: ({}){:?} ", r_val.len(), r_val, s_val.len(), s_val);
+    let mut sigval: Vec<u8> = Vec::with_capacity(64);
+    sigval.extend(r_val);
+    sigval.extend(s_val);
+
+    let auth_t = format!(
+        "{}.{}",
+        content.clone(),
+        base64::encode_config(
+            unsafe { &String::from_utf8_unchecked(sigval) },
+            base64::URL_SAFE_NO_PAD,
+        )
+    );
+
+    Ok(format!(
+        "Authorization: {} t={},k={}",
+        SCHEMA,
+        auth_t,
+        auth_k
+    ))
+}
+
+pub fn verify(auth_token: String) -> Result<HashMap<String, serde_json::Value>, String> {
+
+    //Verify that the auth token string matches for the verification token string
+    let auth_token = parse_auth_token(&mut String::from(auth_token.clone()))
+        .expect("Authorization header is invalid.");
+    let pub_ec_key =
+        Key::from_public_raw(auth_token.k).expect("'k' token is not a valid public key");
+    let pub_key = &match pkey::PKey::from_ec_key(pub_ec_key.0) {
+        Ok(key) => key,
+        Err(err) => return Err(format!("Public Key Generation error: {:?}", err)),
+    };
+    let mut verifier = match Verifier::new(MessageDigest::sha256(), pub_key) {
+        Ok(verifier) => verifier,
+        Err(err) => return Err(format!("Verifier failed to initialize: {:?}", err)),
+    };
+
+    let data = &auth_token.t[0].clone().into_bytes();
+    let verif_sig = base64::decode_config(
+        &auth_token.t[1].clone().into_bytes(),
+        base64::URL_SAFE_NO_PAD,
+    ).expect("Signature failed to decode from base64");
+    verifier.update(data).expect(
+        "Data failed to load into verifier",
+    );
+
+    // Extract the values from the combined raw key.
+    let mut r_val = Vec::with_capacity(32);
+    let mut s_val = Vec::with_capacity(32);
+    r_val.extend(verif_sig[0..32].iter());
+    s_val.extend(verif_sig[32..].iter());
+
+    /* Compose the sequence DER by hand, because the current rust libraries lack this. */
+    // write r & s as asn1
+    // Prefix is the "\x02" + the length. We can cheat here because we know how long the keys are.
+    let mut r_asn = vec![2];
+    let mut s_asn = vec![2];
+    // check if we need to pad for high order byte
+    if r_val[0] > 127 {
+        r_asn.append(&mut vec![33, 0])
```

Comment:
To avoid the extra `vec!` here you can also use `r_asn.extend_from_slice(&[33, 0])` I believe